lib/spack/spack/environment/environment.py

```diff
@@ -422,25 +422,105 @@ def _eval_conditional(string):
     return eval(string, valid_variables)
 
 
-def _is_dev_spec_and_has_changed(spec):
+def _timestamp_changed(spec, _database=None):
     """Check if the passed spec is a dev build and whether it has changed since the
     last installation"""
-    # First check if this is a dev build and in the process already try to get
-    # the dev_path
-    dev_path_var = spec.variants.get("dev_path", None)
-    if not dev_path_var:
-        return False
-
-    # Now we can check whether the code changed since the last installation
-    if not spec.installed:
-        # Not installed -> nothing to compare against
-        return False
-
-    _, record = spack.store.STORE.db.query_by_spec_hash(spec.dag_hash())
+    db = _database or spack.store.STORE.db
+    dev_path_var = spec.variants.get("dev_path")
+    _, record = db.query_by_spec_hash(spec.dag_hash())
     mtime = fs.last_modification_time_recursive(dev_path_var.value)
     return mtime > record.installation_time
 
 
+class GitRepoChangeDetector:
+    """Associates with a Git repository. Uses Git to determine when
+    tracked files in the repository have changed, regardless of whether
+    they have been committed or staged (i.e. with ``git add``).
+
+    This does not look for changes in untracked files, and maintains
+    extra files for bookkeeping in this directory, within a ``.spack``
+    subdirectory (creating it if it does not exist).
+    """
+
+    CHANGED = 0
+    NO_PRIOR = 1
+    NOT_CHANGED = 2
+
+    def __init__(self, dev_path):
+        self.base_dir = pathlib.Path(dev_path)
+        self.git_dir = self.base_dir / ".git"
+        self.spack_state = self.base_dir / ".spack"
+        self.cache_state = self.spack_state / "spackdev-git-hash"
+        self.current_hash = None
+
+    @staticmethod
+    def from_src_dir(dev_path):
+        git_dir = pathlib.Path(dev_path) / ".git"
+        if not git_dir.is_dir():
+            return
+        return GitRepoChangeDetector(dev_path)
+
+    def git_modification_hash(self):
+        """
+        Given a Git repository directory, output a hash that represents the current
+        state of the directory, including all uncommitted changes to tracked files
+        in the repository.
+        """
+        git_index = self.git_dir / "index"
+        tmp_index = self.spack_state / "spack-package-git-index"
```

Comment:
We should use a proper temp file for this -- b/c it's likely to be faster (i.e., in ramdisk) and you won't have to talk to the shared filesystem as much.  So, e.g.:

```python
with tempfile.TemporaryDirectory() as tmp:
    tmp_index = os.path.join(tmp, 'tmp_index')
    # copy index to tmp_index
```

that will automatically create a tmp directory and remove it when you're done.